pandas/core/frame.py

```diff
@@ -5070,14 +5070,38 @@ def notna(self) -> DataFrame:
     def notnull(self) -> DataFrame:
         return ~self.isna()
 
+    @overload
+    # https://github.com/python/mypy/issues/6580
+    # Overloaded function signatures 1 and 2 overlap with incompatible return types
+    def dropna(  # type: ignore[misc]
+        self,
+        axis: Axis = ...,
+        how: str = ...,
+        thresh: Optional[int] = ...,
+        subset: Optional[Union[Hashable, Sequence[Hashable]]] = ...,
+        inplace: Literal[False] = ...,
+    ) -> DataFrame:
+        ...
+
+    @overload
+    def dropna(
+        self,
+        axis: Axis = ...,
+        how: str = ...,
+        thresh: Optional[int] = ...,
+        subset: Optional[Union[Hashable, Sequence[Hashable]]] = ...,
+        inplace: Literal[True] = ...,
+    ) -> None:
+        ...
+
     def dropna(
         self,
         axis: Axis = 0,
         how: str = "any",
-        thresh=None,
-        subset=None,
+        thresh: Optional[int] = None,
+        subset: Optional[Union[Hashable, Sequence[Hashable]]] = None,
         inplace: bool = False,
-    ):
+    ) -> Optional[DataFrame]:
```

Comment:
I see you used #37137 as a example. However the motivation for that PR was to remove an `assert df is not None` from the codebase and the return type of `Optional[DataFrame]` not touched.

Not needed to be done in this PR, but just FYI if the motivation for adding the types is for the public api, then the return type should be the same type as self for subclassed DataFrames.